crates/driver/src/tests/cases/protocol_fees.rs

```diff
@@ -173,18 +243,179 @@ async fn volume_protocol_fee_buy_order() {
+            network_fee: 1.ether().into_wei(),
+        },
+    };
     let test_case = TestCase {
-        order_side: order::Side::Sell,
         fee_policy,
-        order_sell_amount: 50.ether().into_wei(),
-        solver_fee: Some(10.ether().into_wei()),
-        quote_sell_amount: 50.ether().into_wei(),
-        quote_buy_amount: 40.ether().into_wei(),
-        executed: 40.ether().into_wei(),
-        executed_sell_amount: 50.ether().into_wei(),
-        executed_buy_amount: 20.ether().into_wei(),
+        order: Order {
+            amounts: Amounts {
+                sell: 50.ether().into_wei(),
+                // Willing to receive less than quoted (in-market)
+                buy: 40.ether().into_wei(),
+            },
+            side: order::Side::Sell,
+        },
+        execution: Execution {
+            // Receive 10 ETH more than quoted, half of which gets captured by the settlement
+            // contract
+            solver: Amounts {
+                sell: 50.ether().into_wei(),
+                buy: 60.ether().into_wei(),
+            },
+            driver: Amounts {
+                sell: 50.ether().into_wei(),
+                buy: 55.ether().into_wei(),
+            },
+        },
     };
+    protocol_fee_test_case(test_case).await;
+}
 
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_buy_out_of_market_order() {
+    let fee_policy = Policy::PriceImprovement {
+        factor: 0.5,
+        max_volume_factor: 1.0,
+        quote: Quote {
+            sell: 59.ether().into_wei(),
+            buy: 40.ether().into_wei(),
+            network_fee: 1.ether().into_wei(),
+        },
+    };
+    let test_case = TestCase {
+        fee_policy,
+        order: Order {
+            amounts: Amounts {
+                // Willing to sell less than quoted (out-market)
+                sell: 50.ether().into_wei(),
+                buy: 40.ether().into_wei(),
+            },
+            side: order::Side::Buy,
+        },
+        execution: Execution {
+            // Sell 10 ETH less than requested, half of which is kept by the protocol
+            solver: Amounts {
+                sell: 40.ether().into_wei(),
+                buy: 40.ether().into_wei(),
+            },
+            driver: Amounts {
+                sell: 45.ether().into_wei(),
+                buy: 40.ether().into_wei(),
+            },
+        },
+    };
+
+    protocol_fee_test_case(test_case).await;
+}
+
+#[tokio::test]
+#[ignore]
+async fn price_improvement_fee_sell_out_of_market_order() {
+    let fee_policy = Policy::PriceImprovement {
+        factor: 0.5,
+        max_volume_factor: 1.0,
+        quote: Quote {
+            sell: 50.ether().into_wei(),
+            buy: 41.ether().into_wei(),
+            network_fee: 1.ether().into_wei(),
+        },
+    };
+    let test_case = TestCase {
+        fee_policy,
+        order: Order {
+            amounts: Amounts {
+                sell: 50.ether().into_wei(),
```

Comment:
It's somewhat confusing that the net quoted after fee (51) and sell amounts (50) are not the same. Is there a reason for that? Otherwise let's make it as easy as possible to reason about this test please.